lib/manager/components/deployment/CustomFileEditor.js

```diff
@@ -0,0 +1,299 @@
+  }
+
+  /**
+   * Makes sure that a uri or some raw input has been added
+   */
+  _hasContents = (): boolean => {
+    const {model: customFile} = this.state
+    return !!customFile.contents || !!customFile.uri
+  }
+
+  /**
+   * Makes sure that the file is used during either graph building, serving or
+   * both.
+   */
+  _hasSomeUsage = (): boolean => {
+    const {model: customFile} = this.state
+    return customFile.useDuringBuild || customFile.useDuringServe
+  }
+
+  _isEditing= () => this.props.idx === this.props.customFileEditIdx
+
+  _isValidOverall = (): boolean => {
+    return this._fileNameValid() && this._hasSomeUsage() && this._hasContents()
+  }
+
+  _onChangeBuildUse = () => {
+    const { model } = this.state
+    this.setState({
+      model: {
+        ...model,
+        useDuringBuild: !model.useDuringBuild
+      }
+    })
+  }
+
+  _onChangeContents = (evt: SyntheticInputEvent<HTMLInputElement>) => {
+    this.setState({
+      model: {
+        ...this.state.model,
+        contents: evt.target.value
+      }
+    })
+  }
+
+  _onChangeFilename = (evt: SyntheticInputEvent<HTMLInputElement>) => {
+    this.setState({
+      model: {
+        ...this.state.model,
+        filename: evt.target.value
+      }
+    })
+  }
+
+  _onChangeServeUse = () => {
+    const { model } = this.state
+    this.setState({
+      model: {
+        ...model,
+        useDuringServe: !model.useDuringServe
+      }
+    })
+  }
+
+  _onChangeSource = (evt: SyntheticInputEvent<HTMLSelectElement>) => {
+    const model = {...this.state.model}
+    // set variable to make flow happy
+    let newSource
+    if (evt.target.value === 'raw') {
+      model.uri = null
+      newSource = 'raw'
+    } else {
+      model.contents = null
+      newSource = 'uri'
+    }
+    this.setState({
+      fileSource: newSource,
+      model
+    })
+  }
+
+  _onChangeUri = (evt: SyntheticInputEvent<HTMLInputElement>) => {
+    this.setState({
+      model: {
+        ...this.state.model,
+        uri: evt.target.value
+      }
+    })
+  }
+
+  _onCancelEditing = () => {
+    const { customFile, onCancelEditing } = this.props
+    this.setState({
+      fileSource: customFile.contents ? 'raw' : 'uri',
+      model: customFile
+    })
+    onCancelEditing()
+  }
+
+  _onDelete = () => {
+    const { idx, onDelete } = this.props
+    onDelete(idx)
+  }
+
+  _onEdit = () => {
+    const { idx, onEdit } = this.props
+    onEdit(idx)
+  }
+
+  _onSave = () => {
+    const { idx, onSave } = this.props
+    onSave(idx, this.state.model)
+  }
+
+  _renderToolbar = () => {
+    const isEditing = this._isEditing()
+    const canEdit = this._canEdit()
+    const either = !isEditing && !canEdit
+    return (
+      <ButtonToolbar>
+        <Button
+          bsSize='xsmall'
+          disabled={either}
+          onClick={canEdit ? this._onEdit : this._onCancelEditing}
+        >
+          {isEditing ? 'Cancel' : 'Edit'}
+        </Button>
```

Comment:
See my other comment regarding hiding Cancel when creating a new entry.

Reply:
Cancel button hiding logic implemented in 00445ee